src/fundamental/sha1-fundamental.c

```diff
@@ -256,8 +257,10 @@ void sha1_process_bytes(const void *buffer, size_t size, struct sha1_ctx *ctx) {
         if ((j + size) > 63) {
                 memcpy(&ctx->buffer[j], buffer, (i = 64 - j));
                 sha1_do_transform(ctx->state, ctx->buffer);
-                for (; i + 63 < size; i += 64)
-                        sha1_do_transform(ctx->state, (const uint8_t *) buffer + i);
+                for (; i + 63 < size; i += 64) {
+                        memcpy(buffer64, (const uint8_t *) buffer + i, sizeof(buffer64));
+                        sha1_do_transform(ctx->state, buffer64);
```

Comment:
btw, what precisely is the gcc buid error you are seeing here?

Reply:
Something about incompatible pointers, I don't remember exactly. Like sha1_do_transform expects 64 bytes, but GCC 13 can't prove that here

Reply:
hmm, the memcpy seems quite inefficient here though, this is the inner loop afte rall.

there must be a better way, maybe some explicit cast?

Reply:
Well, yeah, it's inefficient. It only happens in CI because of broken GCC 13, I cannot reproduce locally with GCC 14.
And type tetris is very painful if each change takes >5 minutes to test

Reply:
Okay, that's stupid. I can't reproduce locally with gcc-13 either.

Reply:
```
2024-10-11T11:16:26.0180314Z In function ‘sha1_process_bytes’,
2024-10-11T11:16:26.0181459Z     inlined from ‘sha1_finish_ctx’ at ../src/fundamental/sha1-fundamental.c:277:9:
2024-10-11T11:16:26.0183719Z ../src/fundamental/sha1-fundamental.c:260:25: error: ‘sha1_do_transform’ reading 64 bytes from a region of size 1 [-Werror=stringop-overread]
2024-10-11T11:16:26.0185631Z   260 |                         sha1_do_transform(ctx->state, (const uint8_t *) buffer + i);
2024-10-11T11:16:26.0186620Z       |                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2024-10-11T11:16:26.0188410Z ../src/fundamental/sha1-fundamental.c:260:25: note: referencing argument 2 of type ‘const uint8_t[64]’ {aka ‘const unsigned char[64]’}
2024-10-11T11:16:26.0190272Z ../src/fundamental/sha1-fundamental.c: In function ‘sha1_finish_ctx’:
2024-10-11T11:16:26.0191750Z ../src/fundamental/sha1-fundamental.c:125:13: note: in a call to function ‘sha1_do_transform’
2024-10-11T11:16:26.0193361Z   125 | static void sha1_do_transform(uint32_t state[5], const uint8_t buffer[64]) {
2024-10-11T11:16:26.0194303Z       |             ^~~~~~~~~~~~~~~~~
2024-10-11T11:16:26.0194983Z In function ‘sha1_process_bytes’,
2024-10-11T11:16:26.0196021Z     inlined from ‘sha1_finish_ctx’ at ../src/fundamental/sha1-fundamental.c:279:17:
2024-10-11T11:16:26.0198197Z ../src/fundamental/sha1-fundamental.c:260:25: error: ‘sha1_do_transform’ reading 64 bytes from a region of size 1 [-Werror=stringop-overread]
2024-10-11T11:16:26.0200045Z   260 |                         sha1_do_transform(ctx->state, (const uint8_t *) buffer + i);
2024-10-11T11:16:26.0200982Z       |                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2024-10-11T11:16:26.0232955Z ../src/fundamental/sha1-fundamental.c:260:25: note: referencing argument 2 of type ‘const uint8_t[64]’ {aka ‘const unsigned char[64]’}
2024-10-11T11:16:26.0234885Z ../src/fundamental/sha1-fundamental.c: In function ‘sha1_finish_ctx’:
2024-10-11T11:16:26.0236354Z ../src/fundamental/sha1-fundamental.c:125:13: note: in a call to function ‘sha1_do_transform’
2024-10-11T11:16:26.0237742Z   125 | static void sha1_do_transform(uint32_t state[5], const uint8_t buffer[64]) {
2024-10-11T11:16:26.0238674Z       |             ^~~~~~~~~~~~~~~~~
2024-10-11T11:16:26.0239287Z cc1: all warnings being treated as errors
```

It takes almost 30 minutes to fail in CI and it never fails locally with 
```
$ gcc-13 --version
gcc-13 (GCC) 13.3.0
```
I can't properly fix something that can't easily reproduce locally.

Reply:
It's not very clear from the log, but the full configuration of that build test should be the following, to reproduce it:

```
AR=gcc-ar-13
CC=gcc-13
CC_LD=mold
CXX=g++-13
CXX_LD=mold
meson setup -Dtests=unsafe -Dslow-tests=true -Dfuzz-tests=true --werror -Dnobody-group=nogroup -Dcryptolib=openssl -Ddebug=false --optimization=3 -Db_lto=false -Dtpm2=disabled -Dlibfido2=disabled -Dp11kit=disabled build
```

Reply:
so afaics gcc is just wrong. Might be ok in this case to tape over this, i.e. add to src/fundamental/macro-fundamental.h

```c
#define DISABLE_WARNING_STRINGOP_OVERREAD                               \
        _Pragma("GCC diagnostic push");                                 \
        _Pragma("GCC diagnostic ignored \"-Wstringop-overread\"")
```

And then make use of that here. see how this used in the other DISABLE_WARNING_xyz cases.